pkg/buildexpression/parser/parser.go

```diff
@@ -0,0 +1,541 @@
+	for {
+		if p.tok != L_CURL && !isString(p.tok) && p.tok != COMMA {
+			break
+		}
+
+		if p.tok == COMMA {
+			node.AddChild(p.newNode(NodeComma))
+
+			p.next()
+			continue
+		}
+
+		var err error
+		var msg string
+		switch p.tok {
+		case L_CURL:
+			err = p.parseListObject(node)
+			msg = "Failed to parse list object"
+		case STRING:
+			err = p.parseListString(node)
+			msg = "Failed to parse list string"
+		}
+		if err != nil {
+			return errs.Wrap(err, msg)
+		}
+	}
+
+	return nil
+}
+
+func (p *Parser) parseListString(node *Node) error {
+	elementNode := p.newNode(NodeListElement)
+	node.AddChild(elementNode)
+
+	return p.parseString(elementNode)
+}
+
+func (p *Parser) parseListObject(node *Node) error {
+	elementNode := p.newNode(NodeListElement)
+	node.AddChild(elementNode)
+
+	objectNode := p.newNode(NodeObject)
+	elementNode.AddChild(objectNode)
+
+	err := p.expectToken(L_CURL, objectNode, NodeLeftCurlyBracket)
+	if err != nil {
+		return errs.Wrap(err, "Failed to scan")
+	}
+
+	for {
+		if p.tok != IDENTIFIER {
+			break
+		}
+
+		err := p.parseObjectAttribute(objectNode)
+		if err != nil {
+			return errs.Wrap(err, "Failed to parse object attribute")
+		}
+
+		if p.tok == COMMA {
+			objectNode.AddChild(p.newNode(NodeComma))
+			p.next()
+		}
+	}
+
+	return p.expectToken(R_CURL, objectNode, NodeRightCurlyBracket)
+}
+
+func (p *Parser) parseObjectAttribute(node *Node) error {
+	if p.tok != IDENTIFIER {
+		return errs.New("Expected identifier")
+	}
+
+	err := p.parseIdentifier(node)
+	if err != nil {
+		return errs.Wrap(err, "Failed to parse identifier")
+	}
+
+	err = p.expectToken(COLON, node, NodeColon)
+	if err != nil {
+		return errs.Wrap(err, "Expect failed")
+	}
+
+	var msg string
+	switch p.tok {
+	case STRING:
+		err = p.parseString(node)
+		msg = "Failed to parse string"
+	case L_CURL:
+		err = p.parseListObject(node)
+		msg = "Failed to parse list object"
+	default:
+		err = p.parseList(node)
+		msg = "Failed to parse list"
+	}
+	if err != nil {
+		return errs.Wrap(err, msg)
+	}
+
+	return nil
+}
+
+func (p *Parser) parseString(node *Node) error {
+	if p.tok != STRING {
+		return errs.New("Expected string")
+	}
+	node.AddChild(p.newNode(NodeString))
+
+	p.next()
+	return nil
+}
+
+func (p *Parser) parseIn(node *Node) error {
+	if p.tok != IN {
+		return errs.New("Expected in")
+	}
+	node.AddChild(p.newNode(NodeIn))
+
+	p.next()
+
+	err := p.expectToken(COLON, node, NodeColon)
+	if err != nil {
+		return errs.Wrap(err, "Expect failed")
+	}
+
+	var msg string
+	switch p.tok {
+	case STRING:
+		err = p.parseString(node)
+		msg = "Failed to parse string"
+	case IDENTIFIER:
+		err = p.parseApplication(node)
+		msg = "Failed to parse application"
+	default:
+		return errs.New("Expected string or identifier, got: %s", p.lit)
+	}
+	if err != nil {
+		return errs.Wrap(err, msg)
+	}
+
+	return nil
+}
+
+func (p *Parser) isAssignment() bool {
+	return p.tok == IDENTIFIER && p.peek() == COLON && p.tok != IN
```

Comment:
Is it possible for `p.tok == IDENTIFIER` and `p.tok == IN` to both be true at once? If not, then you can remove the `p.tok != IN` test.